Lib/test/string_tests.py

```diff
@@ -318,7 +318,7 @@ def test_rindex(self):
     def test_find_periodic_pattern(self):
         """Cover the special path for periodic patterns."""
         def reference_find(p, s):
-            for i in range(len(s)):
+            for i in range(len(s)+1):
                 if s.startswith(p, i):
                     return i
```

Comment:
I verified that `''.startswith('', 0) == True`, so that `reference_find('', '') == 0`.  How convinced are you that the `+1` cannot create a failure in another corner case?

Comment: this would be easier for me to read in relation to the below if `s` were `text` or vice versa.

Reply:
I re-named `s -> text`, and re-ordered the arguments to match those of `str.find()`.

> How convinced are you that the +1 cannot create a failure in another corner case?

I am fairly confident. You could just as well reference_find as
```
    def reference_find(text, p):
        substrings = [text[i:i+len(p)] for i in range(len(text) - len(p) + 1)]
        try:
            return substrings.index(p)
        except ValueError:
            return -1
```

It doesn't matter whether you use `range(len(text) - len(p) + 1)` or just `range(len(text) + 1)`. The last possible match occurs when `i = len(text) - len(p)`, and for anything after that, the suffix `text[i:]` is too short, so none of those will equal `p`. Perhaps this addition of `range(len(text) - len(p) + 1)` makes this a little bit clearer why the +1 is required. The range `range(len(text))` is only shorter than `range(len(text) - len(p) + 1)` when `p == ''`, so that is the only case where reference_find could is incorrect. Since `text.find('')` should always return `-1`, we just need to fix it so that the range always includes starting at 0.

Using `text.startswith(p, i)` is just an optimization: slicing `text[i:i+len(p)]` at every step and therefore allocating a new string at every step is wasteful, especially since the mismatch usually occurs in the first character or two.

Just to be totally sure, I verified that both of these implementations pass all of the other `find()` tests in `string_tests.py`:

<details>

```python
from test.string_tests import BaseTest
import unittest


class ReferenceFindStr1(str):
    def find(self, p, *args):
        if args:
            # handle start, end
            return str(self).find(str(p), *args)
        text = self
        substrings = [text[i:i+len(p)] for i in range(len(text) - len(p) + 1)]
        try:
            return substrings.index(p)
        except ValueError:
            return -1

class ReferenceFindStr2(str):
    def find(self, p, *args):
        if args:
            # handle start, end
            return str(self).find(str(p), *args)
        text = self
        for i in range(len(text) - len(p) + 1):
            if text.startswith(p, i):
                return i
        return -1


class TestReferenceFindBase(BaseTest):
    def test_adaptive_find(self):
        raise unittest.SkipTest("too slow")
    
    def test_find_many_lengths(self):
        raise unittest.SkipTest("too slow")


class TestReferenceFind1(TestReferenceFindBase, unittest.TestCase):
    type2test = ReferenceFindStr1

class TestReferenceFind2(TestReferenceFindBase, unittest.TestCase):
    type2test = ReferenceFindStr2


if __name__ == '__main__':
    unittest.main()
```
</details>

Reply:
ok